src/components/tabs/ScatterPlotTab.tsx

```diff
@@ -386,6 +392,131 @@ export default memo(function ScatterPlotTab(props: ScatterPlotTabProps): ReactEl
     );
   };
 
+  /**
+   * Applies coloring to point traces in a scatterplot. Does this by splitting the data into multiple traces each with a solid
+   * color, which is much faster than using Plotly's native color ramping. Also enforces a maximum number of points
+   * per trace, which significantly speeds up Plotly renders.
+   *
+   * @param xData
+   * @param yData
+   * @param objectIds
+   * @param trackIds
+   * @param markerConfig Additional marker configuration to apply to all points. By default,
+   * markers are size 4.
+   * @param {Color | undefined} overrideColor When defined, uses a base color for all points, instead of
+   * calculating based on the color ramp or palette.
+   */
+  const colorizeScatterplotPoints = (
+    xData: DataArray,
+    yData: DataArray,
+    objectIds: number[],
+    trackIds: number[],
+    markerConfig: Partial<PlotMarker> = {},
+    overrideColor?: Color
+  ): Partial<PlotData>[] => {
+    if (selectedFeatureName === null || dataset === null || !xAxisFeatureName || !yAxisFeatureName) {
+      return [];
+    }
+
+    const featureData = dataset.getFeatureData(selectedFeatureName);
+
+    const traceDataBuckets: TraceData[] = [];
+    if (!featureData || markerConfig.color || overrideColor) {
+      // Do no coloring! Keep all points in the same bucket.
```

Comment:
Would we get a speedup if we binned anyway?? just wondering.

Reply:
These points will get binned below in an extra step! I'll update this comment.